[CSS] Improve performance of visitAllSimpleSelectors()

mdubet opened the pull request:
#### c8893f6f23dfc9d9cdb32a715457c77a70d9102f
<pre>
[CSS] Improve performance of visitAllSimpleSelectors()
<a href="https://bugs.webkit.org/show_bug.cgi?id=259828">https://bugs.webkit.org/show_bug.cgi?id=259828</a>
rdar://113397883

Reviewed by Brent Fulgham.

This patch makes the visitAllSimpleSelectors() function more optimized by:
 - being iterative instead of recursive
 - allowing the functor argument to early return.

It improves performance on really long selectors.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::visitAllSimpleSelectors const):
(WebCore::CSSSelector::resolveNestingParentSelectors):
(WebCore::CSSSelector::replaceNestingParentByPseudoClassScope):
(WebCore::CSSSelector::hasExplicitNestingParent const):
* Source/WebCore/css/CSSSelector.h:

Canonical link: <a href="https://commits.webkit.org/266651@main">https://commits.webkit.org/266651@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04d2563843660cfe47f4632135a1c33640063f44

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14356 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14667 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15009 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16096 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13591 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14481 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17181 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14745 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16255 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14535 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15079 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12181 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16815 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12364 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12941 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/19950 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13441 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13105 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16316 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13661 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11505 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12946 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3475 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17284 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13506 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rreno commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -979,23 +979,28 @@ CSSSelector::CSSSelector(const CSSSelector& other)
-    // Visit the next simple selector recursively
-    if (auto next = tagHistory())
-        next->visitAllSimpleSelectors(apply); 
+        // Visit the next simple selector recursively
```
Stale comment - we're not recursing anymore!

rreno commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -979,23 +979,28 @@ CSSSelector::CSSSelector(const CSSSelector& other)
+        if (auto selectorList = current->selectorList()) {
+            auto next = selectorList->first();
+            while (next) {
+                if (next->visitAllSimpleSelectors(apply))
```
Should this be converted to iteration too?

mdubet replied:
Ideally yes, but it's more complex (need a queue?) and not sure it's worth it. Maybe for another PR 😅

mdubet replied:
Fixed ✅ (with a work queue indeed)

brentfulgham reviewed (approved):
It might help if your ChangeLog said that it "makes it possible to early return" rather than "makes it early return", since it allows the functor to determine which behavior you get.

mdubet committed:
[CSS] Improve performance of visitAllSimpleSelectors()
https://bugs.webkit.org/show_bug.cgi?id=259828
rdar://113397883

Reviewed by Brent Fulgham.

This patch makes the visitAllSimpleSelectors() function more optimized by:
 - being iterative instead of recursive
 - allowing the functor argument to early return.

It improves performance on really long selectors.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::visitAllSimpleSelectors const):
(WebCore::CSSSelector::resolveNestingParentSelectors):
(WebCore::CSSSelector::replaceNestingParentByPseudoClassScope):
(WebCore::CSSSelector::hasExplicitNestingParent const):
* Source/WebCore/css/CSSSelector.h:

Canonical link: https://commits.webkit.org/266651@main
